x-pack/filebeat/input/awss3/sqs.go

```diff
@@ -41,29 +46,112 @@ func newSQSReader(log *logp.Logger, metrics *inputMetrics, sqs sqsAPI, maxMessag
 		maxMessagesInflight: maxMessagesInflight,
 		workerSem:           awscommon.NewSem(maxMessagesInflight),
 		sqs:                 sqs,
+		pipeline:            pipeline,
 		msgHandler:          msgHandler,
 		log:                 log,
 		metrics:             metrics,
 	}
 }
 
+type processingData struct {
+	msg   types.Message
+	start time.Time
+}
+
 func (r *sqsReader) Receive(ctx context.Context) error {
-	// This loop tries to keep the workers busy as much as possible while
-	// honoring the max message cap as opposed to a simpler loop that receives
-	// N messages, waits for them all to finish, then requests N more messages.
-	var workerWg sync.WaitGroup
-	for ctx.Err() == nil {
-		// Determine how many SQS workers are available.
-		workers, err := r.workerSem.AcquireContext(r.maxMessagesInflight, ctx)
+	workersWg := new(sync.WaitGroup)
+	workersWg.Add(r.maxMessagesInflight)
+	workersChan := make(chan processingData, r.maxMessagesInflight)
+
+	deletionWg := new(sync.WaitGroup)
+	deletionWaiter := new(atomic.Bool)
+
+	var clientsMutex sync.Mutex
+	clients := make(map[uint64]beat.Client, r.maxMessagesInflight)
+
+	// Start a fixed amount of goroutines that will process all the SQS messages sent to the workersChan asynchronously.
+	for i := 0; i < r.maxMessagesInflight; i++ {
+		id := r.metrics.beginSQSWorker()
+
+		// Create a pipeline client scoped to this goroutine.
+		client, err := r.pipeline.ConnectWith(beat.ClientConfig{
+			EventListener: NewEventACKHandler(),
+			Processing: beat.ProcessingConfig{
+				// This input only produces events with basic types so normalization
+				// is not required.
+				EventNormalization: boolPtr(false),
+			},
+		})
+
+		clientsMutex.Lock()
+		clients[id] = client
+		clientsMutex.Unlock()
```

Comment:
Is concurrent access to the map actually possible here?

Reply:
no indeed